src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,324 @@
+  // Return true if token allowance is greater than input amount
+  return allowance.gt(BigNumber.from(inputAmount.quotient.toString()));
+};
+
+export class FallbackTenderlySimulator implements ISimulator {
+  private provider: JsonRpcProvider;
+  private tenderlySimulator: TenderlySimulator;
+  v3PoolProvider: IV3PoolProvider;
+  v2PoolProvider: IV2PoolProvider;
+
+  constructor(
+    tenderlyBaseUrl: string,
+    tenderlyUser: string,
+    tenderlyProject: string,
+    tenderlyAccessKey: string,
+    provider: JsonRpcProvider,
+    v2PoolProvider: IV2PoolProvider,
+    v3PoolProvider: IV3PoolProvider,
+    tenderlySimulator?: TenderlySimulator
+  ) {
+    this.tenderlySimulator =
+      tenderlySimulator ??
+      new TenderlySimulator(
+        tenderlyBaseUrl,
+        tenderlyUser,
+        tenderlyProject,
+        tenderlyAccessKey,
+        v2PoolProvider,
+        v3PoolProvider
+      );
+    this.provider = provider;
+    this.v2PoolProvider = v2PoolProvider;
+    this.v3PoolProvider = v3PoolProvider;
+  }
+
+  private async ethEstimateGas(
+    fromAddress: string,
+    route: SwapRoute,
+    l2GasData?: ArbitrumGasData | OptimismGasData
+  ): Promise<SwapRoute> {
+    const currencyIn = route.trade.inputAmount.currency;
+    const router = SwapRouter02__factory.connect(
+      SWAP_ROUTER_ADDRESS,
+      this.provider
+    );
+    const estimatedGasUsed: BigNumber = await router.estimateGas[
+      'multicall(bytes[])'
+    ]([route.methodParameters!.calldata], {
+      from: fromAddress,
+      value: BigNumber.from(
+        currencyIn.isNative ? route.methodParameters!.value : '0'
+      ),
+    });
+    const {
+      estimatedGasUsedUSD,
+      estimatedGasUsedQuoteToken,
+      quoteGasAdjusted,
+    } = await calculateGasUsed(
+      route.quote.currency.chainId,
+      route,
+      estimatedGasUsed,
+      this.v2PoolProvider,
+      this.v3PoolProvider,
+      l2GasData
+    );
+    return initSwapRouteFromExisting(
+      route,
+      this.v2PoolProvider,
+      this.v3PoolProvider,
+      quoteGasAdjusted,
+      estimatedGasUsed,
+      estimatedGasUsedQuoteToken,
+      estimatedGasUsedUSD
+    );
+  }
+
+  public async simulateTransaction(
+    fromAddress: string,
+    swapRoute: SwapRoute,
+    l2GasData?: ArbitrumGasData | OptimismGasData
+  ): Promise<SwapRoute> {
+    // Make call to eth estimate gas if possible
+    // For erc20s, we must check if the token allowance is sufficient
+    const inputAmount = swapRoute.trade.inputAmount;
+    if (
+      inputAmount.currency.isNative ||
+      (await checkTokenApproved(fromAddress, inputAmount, this.provider))
+    ) {
+      try {
+        const swapRouteWithGasEstimate = await this.ethEstimateGas(
+          fromAddress,
+          swapRoute,
+          l2GasData
+        );
+        return swapRouteWithGasEstimate;
+      } catch (err) {
+        log.info({ err: err }, 'Error calling eth estimate gas!');
+        return { ...swapRoute, simulationError: false };
```

Comment:
Should be `true`. The simulator class failed to produce an accurate simulated gas estimate and didn't update the swapRoute object, so the flag needs to be set to true